components/eventing-controller/reconciler/backend/reconciler.go

```diff
@@ -0,0 +1,697 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/kyma-project/kyma/components/eventing-controller/pkg/object"
+	"github.com/pkg/errors"
+
+	appsv1 "k8s.io/api/apps/v1"
+
+	"github.com/go-logr/logr"
+	eventingv1alpha1 "github.com/kyma-project/kyma/components/eventing-controller/api/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	BEBBackendSecretLabelKey   = "kyma-project.io/eventing-backend"
+	BEBBackendSecretLabelValue = "beb"
+
+	BackendCRLabelKey   = "kyma-project.io/eventing"
+	BackendCRLabelValue = "backend"
+
+	AppLabelValue             = deployment.PublisherName
+	PublisherSecretEMSHostKey = "ems-publish-host"
+
+	TokenEndpointFormat             = "%s?grant_type=%s&response_type=token"
+	NamespacePrefix                 = "/"
+	BEBPublishEndpointForSubscriber = "/sap/ems/v1"
+	BEBPublishEndpointForPublisher  = "/sap/ems/v1/events"
+)
+
+type Reconciler struct {
+	ctx                  context.Context
+	natsCommander        commander.Commander
+	natsCommanderStarted bool
+	bebCommander         commander.Commander
+	bebCommanderStarted  bool
+	client.Client
+	// TODO: Do we need to explicitly pass and use a cache here? The default client that we get from manager
+	//  already uses a cache internally (check manager.DefaultNewClient)
+	cache.Cache
+	Log    logr.Logger
+	record record.EventRecorder
+	cfg    env.BackendConfig
+}
+
+func NewReconciler(ctx context.Context, natsCommander, bebCommander commander.Commander, client client.Client, cache cache.Cache, log logr.Logger, recorder record.EventRecorder) *Reconciler {
+	cfg := env.GetBackendConfig()
+	return &Reconciler{
+		ctx:           ctx,
+		natsCommander: natsCommander,
+		bebCommander:  bebCommander,
+		Client:        client,
+		Cache:         cache,
+		Log:           log,
+		record:        recorder,
+		cfg:           cfg,
+	}
+}
+
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;update;patch;create;delete
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;update;patch;create;delete
+// +kubebuilder:rbac:groups=eventing.kyma-project.io,resources=eventingbackends,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=eventing.kyma-project.io,resources=eventingbackends/status,verbs=get;update;patch
+
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	var secretList v1.SecretList
+
+	if err := r.Cache.List(ctx, &secretList, client.MatchingLabels{
+		BEBBackendSecretLabelKey: BEBBackendSecretLabelValue,
+	}); err != nil {
+		return ctrl.Result{}, err
+	}
+	r.Log.Info("Found secrets with BEB label", "count", len(secretList.Items))
```

Comment:
If the system is working as expected (no switch), the log file is full with the same set of log messages:
```
...
{"level":"info","ts":1621493705.130021,"logger":"reconciler.backend","msg":"Found secrets with BEB label","count":0}
{"level":"info","ts":1621493705.1300337,"logger":"reconciler.backend","msg":"Reconciling with backend as NATS"}
{"level":"info","ts":1621493705.1300404,"logger":"reconciler.backend","msg":"Found existing backend CR"}
{"level":"info","ts":1621493705.1300511,"logger":"reconciler.backend","msg":"No need to update existing backend CR"}
{"level":"info","ts":1621493705.1302183,"logger":"reconciler.backend","msg":"Created/updated backend CR"}
{"level":"info","ts":1621493705.130247,"logger":"reconciler.backend","msg":"trying to create/update eventing publisher proxy..."}
{"level":"info","ts":1621493705.1302927,"logger":"reconciler.backend","msg":"No need to update publisher proxy"}
{"level":"info","ts":1621493705.1302993,"logger":"reconciler.backend","msg":"Created/updated publisher proxy"}
{"level":"info","ts":1621493705.130316,"logger":"reconciler.backend","msg":"No need to update backend CR status"}
{"level":"info","ts":1621493705.1303806,"logger":"reconciler.backend","msg":"Found secrets with BEB label","count":0}
{"level":"info","ts":1621493705.130397,"logger":"reconciler.backend","msg":"Reconciling with backend as NATS"}
{"level":"info","ts":1621493705.130408,"logger":"reconciler.backend","msg":"Found existing backend CR"}
{"level":"info","ts":1621493705.130419,"logger":"reconciler.backend","msg":"No need to update existing backend CR"}
{"level":"info","ts":1621493705.1304278,"logger":"reconciler.backend","msg":"Created/updated backend CR"}
{"level":"info","ts":1621493705.1304355,"logger":"reconciler.backend","msg":"trying to create/update eventing publisher proxy..."}
{"level":"info","ts":1621493705.1304693,"logger":"reconciler.backend","msg":"No need to update publisher proxy"}
{"level":"info","ts":1621493705.1304784,"logger":"reconciler.backend","msg":"Created/updated publisher proxy"}
{"level":"info","ts":1621493705.1304882,"logger":"reconciler.backend","msg":"No need to update backend CR status"}
...
```

Reply:
Will be solved by: https://github.com/kyma-project/kyma/issues/11321